cluster_management/src/main/java/com/pinterest/rocksplicator/CurrentStatesConfigGenerator.java

```diff
@@ -0,0 +1,505 @@
+            disabledHosts.add(helixHostName);
+            continue;
+          }
+
+          if (!instanceConfig.getInstanceEnabled()
+              || instanceConfig.containsTag("disabled")
+              || !liveInstance.isValid()) {
+            disabledHosts.add(helixHostName);
+            continue;
+          }
+
+          existingHosts.add(helixHostName);
+
+          String state = entry.getValue();
+          if (!ExternalViewUtils.isServing(state)) {
+            continue;
+          }
+
+          String hostWithDomain = getHostWithDomain(helixHostName, instanceConfig);
+          List<String> partitionList = hostToPartitionList.get(hostWithDomain);
+          if (partitionList == null) {
+            partitionList = new ArrayList<String>();
+            hostToPartitionList.put(hostWithDomain, partitionList);
+          }
+          partitionList.add(partitionNumber + ExternalViewUtils.getShortHandState(state));
+        }
+      }
+
+      // Add host to partition list map to the resource config
+      for (Map.Entry<String, List<String>> entry : hostToPartitionList.entrySet()) {
+        JSONArray jsonArray = new JSONArray();
+        for (String p : entry.getValue()) {
+          jsonArray.add(p);
+        }
+
+        resourceConfig.put(entry.getKey(), jsonArray);
+      }
+
+      // add the resource config to the cluster config
+      jsonClusterShardMap.put(resource, resourceConfig);
+    }
+
+    hostToHostWithDomain.keySet().retainAll(existingHosts);
+
+    /**
+     * Finally publish the shard_map in json_format to multiple configured publishers.
+     */
+    shardMapPublisher.publish(
+        externalViewsToProcess.stream().map(e -> e.getResourceName()).collect(Collectors.toSet()),
+        externalViewsToProcess,
+        jsonClusterShardMap);
+
+    long shardPostingTimeMillis = System.currentTimeMillis();
+
+    stopwatch.stop();
+    long elapsedMs = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+    this.monitor.reportConfigGeneratorLatency(elapsedMs);
+
+    // All side-effects must be done, after the shard-map is posted...
+    if (externalViewLeaderEventLogger != null) {
+      LOG.error("Processing ExternalViews for LeaderEventsLogger");
+      externalViewLeaderEventLogger.process(
+          externalViewsToProcess, disabledHosts, generationStartTimeMillis, shardPostingTimeMillis);
+    }
+  }
+
+  private ExternalView createExternalView(IdealState idealState, RoutingTableSnapshot snapshot) {
+    // First number of partitions
+    int numPartitions = idealState.getNumPartitions();
+    String stateModel = idealState.getStateModelDefRef();
+    List<String> states = ImmutableList.of();
+    if (stateModel.equals("LeaderFollower")) {
+      states = ImmutableList.of("LEADER", "FOLLOWER");
+    } else if (stateModel.equals("MasterSlave")) {
+      states = ImmutableList.of("MASTER", "SLAVE");
+    } else if (stateModel.equals("OnlineOffline")) {
+      states = ImmutableList.of("ONLINE");
+    }
```

Comment:
Move this to Utils class (rename ExternalViewUtils to. something common for helix related functions)